### PR TITLE
Make NSObject conform to CVarArg

### PR DESCRIPTION
**Pull request by Kyle-Ye**
## Summary

- Make `NSObject` conform to `CVarArg` so Foundation objects can be passed to `%@` formatting on non-Darwin platforms.
- Route `NSString(format:)` through `_CFStringCreateWithFormatAndArgumentsAux` with a Foundation description callback, matching the CoreFoundation hook intended for `descriptionWithLocale:` / `description`.
- Remove the narrower `NSAttributedString`-only conformance and cover ordinary `NSObject`, `NSAttributedString`, and locale-aware formatting in `TestNSString`.

## Details

`NSObject: CVarArg` only supplies the object pointer to the varargs list. The important behavioral change is that `NSString(format:)` now provides CoreFoundation with a `copyDescFunc`, so `%@` formatting asks Foundation for an object description instead of falling back to treating every object pointer as a CF runtime object.

For known Foundation collection/number/date types, the callback uses the existing locale-aware description entry points. Other `NSObject` subclasses use `object.description`.

### CODE REVIEW

File: Sources/Foundation/NSString.swift [hunk 1]
```diff
@@ -1478,6 +1464,66 @@ extension NSString {
...
+        formatOptions = nil
+    }
+
+    return _CFStringCreateWithFormatAndArgumentsAux(kCFAllocatorSystemDefault, _NSCopyFormattingDescription, formatOptions, format, arguments)
+}
+
+private func _NSCopyFormattingDescription(_ object: UnsafeMutableRawPointer, _ formatOptions: UnsafeRawPointer) -> Unmanaged<CFString> {
+    let value = Unmanaged<AnyObject>.fromOpaque(object).takeUnretainedValue()
+    if let nsObject = value as? NSObject {
+        return Unmanaged.passRetained(_NSFormattingDescription(nsObject, locale: _NSFormattingLocale(formatOptions))._cfObject)
```
Comment (jmschonfeld) [inline Sources/Foundation/NSString.swift]: Instead of the `as? NSObject` and various `as NSX` casts, instead can we perform a selector check here? Ideally this code would look something like the following pseudocode:

```
if value responds to selector "descriptionWithLocale:" {
    value.description(withLocale: formatOptions)
} else if value responds to selector "description" {
    value.description
} else {
    "*nil description*"
}
```

File: Sources/Foundation/NSString.swift [hunk 2]
```diff
@@ -1478,6 +1464,66 @@ extension NSString {
...
+    }
+
+    let cfObject = unsafeBitCast(object, to: CFTypeRef.self)
+    return Unmanaged.passRetained(CFCopyDescription(cfObject))
```
Comment (jmschonfeld) [inline Sources/Foundation/NSString.swift]: I think we can remove this path and just assume everything here will be an `NSObject`. `CFTypeRef`s don't conform to `CVarArg` so we know one can never be provided as an argument when using the NSString formatting functions (and CF functions that do accept the CF types will still work because this up-call isn't provided)

File: Tests/Foundation/TestNSString.swift [hunk 3]
```diff
@@ -862,6 +867,26 @@ class TestNSString: LoopbackServerTest {
         }
     }
 
+    func test_initializeWithFormatNSAttributedStringCVarArg() {
+        let value = NSAttributedString(string: "hello")
+        let result = NSString(format: "%@", value)
+        XCTAssertEqual(result, "hello{\n}")
+    }
+
+    func test_initializeWithFormatNSObjectCVarArg() {
```
Comment (jmschonfeld) [inline Tests/Foundation/TestNSString.swift:876]: Would you mind adding a test here that passes `nil` as an object just to make sure optional handling works ok? I believe it does, but I don't know if we have a test for it yet.

File: Sources/Foundation/NSString.swift (hunk 1 above)
Comment (jmschonfeld) [reply Sources/Foundation/NSString.swift]: Oh, I suppose without the Objective-C runtime, do we have a way to do a responds-to-selector check? Perhaps not...

File: Tests/Foundation/TestNSString.swift (hunk 3 above)
Comment (Kyle-Ye) [reply Tests/Foundation/TestNSString.swift:876]: No problem

File: Sources/Foundation/NSString.swift (hunk 2 above)
Comment (Kyle-Ye) [reply Sources/Foundation/NSString.swift]: Got it.

File: Sources/Foundation/NSString.swift (hunk 1 above)
Comment (Kyle-Ye) [reply Sources/Foundation/NSString.swift]: Yeah. That's issue here. Unless we add a new internal protocol and make those types conform to the new protocol we added. But the code is not less than the current status.

File: Tests/Foundation/TestNSString.swift (hunk 3 above)
Comment (Kyle-Ye) [reply Tests/Foundation/TestNSString.swift:876]: `NSObject?` does not conform to CVarArg too on macOS. Do you mean something like this?

```swift
func test_initializeWithFormatNilObjectCVarArg() {
    let value = unsafeBitCast(0, to: NSObject.self)
    let result = NSString(format: "%@", value)
    XCTAssertEqual(result, "(null)")
}
```

File: Sources/Foundation/NSString.swift (hunk 1 above)
Comment (jmschonfeld) [reply Sources/Foundation/NSString.swift]: Ok then this seems fine to me

File: Tests/Foundation/TestNSString.swift (hunk 3 above)
Comment (jmschonfeld) [reply Tests/Foundation/TestNSString.swift:876]: Oh you're right! From ObjC/C callers you can provide null objects, but Swift prevents it since `Optional` isn't a `CVarArg`, even on macOS. Ok in that case no need to test it since it's not valid to do anyways

**Review by jmschonfeld — APPROVED** [review]

**Comment by jmschonfeld** [comment]
@swift-ci please test

**Comment by jmschonfeld** [comment]
Sorry I didn't catch that you had committed the `test_initializeWithFormatNilObjectCVarArg` - looks like the CI failures are due to that test

File: Tests/Foundation/TestNSString.swift (hunk 3 above)
Comment (Kyle-Ye) [reply Tests/Foundation/TestNSString.swift:876]: Yes. I'll remove this and trigger test again.

**Comment by jmschonfeld** [comment]
@swift-ci please test

**Comment by jmschonfeld** [comment]
@swift-ci please test Windows platform

**Review by jmschonfeld — APPROVED** [review]

**Comment by Kyle-Ye** [comment]
@swift-ci please test Linux platform
